spec-0007/index.md

```diff
@@ -0,0 +1,183 @@
+
+### Constraints
+
+NumPy, SciPy, scikit-learn, scikit-image, and NetworkX all implement pseudo-random seeding in slightly different ways.
+Common keyword arguments include `random_state` and `seed`.
+In practice, the seed is unfortunately also often controlled using `np.random.seed`.
+
+## Implementation
+
+<!--
+Discuss how this would be implemented.
+-->
+
+Legacy behavior in packages such as scikit-learn (`sklearn.utils.check_random_state`) typically handle `None` (use the global seed state), an int (convert to `RandomState`), or `RandomState` object.
+
+Two strong motivations for moving over to `Generator`s are:
+
+(1) they avoid naïve seeding strategies, such as using successive integers, via the underlying [SeedSequence](https://numpy.org/doc/stable/reference/random/parallel.html#seedsequence-spawning);
+(2) they avoid using global state (from `np.random.mtrand._rand`).
+
+Our recommendation here is a deprecation strategy which does not in _all_ cases adhere to the Hinsen[^hinsen] principle.
+
+The [deprecation strategy](https://github.com/scientific-python/specs/pull/180#issuecomment-1515248009) is:
+
+1. Accept both `rng` and `random_state` keyword arguments.
+2. If `rng=None`, handle `random_state` as in legacy behavior (see above), except use a compatible Generator instead of RandomState.
+   A DeprecationWarning is raised to warn about a future change in behavior.
+3. After <X time>, use only `rng`, seeding with `default_rng(rng)`.
+   Raise an error if `random_state` is provided.
+4. At a time of the library's choosing, remove any machinery related to `random_state`.
+
+### Impact
+
+The following users will be affected:
+
+1. Those who use `np.random.seed`. The proposal will do away with that global seeding mechanism, meaning that code that relies on it will, after a certain deprecation period, start seeing a different stream of random numbers than before. To ensure that this does not go unnoticed, the library should raise a `FutureWarning` if `np.random.seed` was called earlier (we show how to do that further down).
+
+   Such code will, in effect, go from being seeded to being unseeded.
+   To avoid that from happening, the code will have to be modified to pass in explicitly an `rng` argument on each function call.
+
+2. Those who do not seed. Their code will, after the deprecation period, use the newly proposed default. Since they were already not requesting repeatable sequences, and since the underlying _distributions_ of pseudo-random numbers did not change, they should be unaffected.
+
+3. Users of `random_state=...`. Support for the `random_state` argument may be dropped eventually, but meanwhile we can provide clear guidance, via deprecation warnings and documentation, on how to migrate to the new `rng` keyword.
+
+[^hinsen]: The Hinsen principle states, loosely, that code should, whether executed now or in the future, return the same result, or raise an error.
+
+### Code
+
```

Comment:
Is there a recommended wording for how the `rng` parameter should be documented?

Reply:
I do recommend emphasizing in each docstring that we expect `Generator`s or `None` (and a quick word about what `None` means) but then link to one common place in your docs that documents all of the rarer cases like static seeds and legacy `RandomState`s. If you use type annotations, consider defining a type that incorporates all the possibilities accepted by `np.random.default_rng()` and documenting that type definition as the place to link those docstrings to.

Reply:
I thought it might be useful to add that type annotation as part of this spec. However, NumPy type annotates with hidden types (`numpy._typing`). This feels like a mistake?

E.g., I tried:

```
import numpy as np
import numpy.typing


type Seed = int | np.typing.ArrayLike | np.random.SeedSequence | np.random.BitGenerator | np.random.Generator | None


def foo(rng : Seed = None):
    rng = np.random.default_rng(rng)

    return rng.random((3, 3))
```

But then mypy (main) complains:

```
$ mypy --enable-incomplete-feature=NewGenericSyntax foo.py
foo.py:9: error: Argument 1 to "default_rng" has incompatible type "int | Buffer | _SupportsArray[dtype[Any]] | _NestedSequence[_SupportsArray[dtype[Any]]] | bool | int | float | complex | str | bytes | _NestedSequence[bool | int | float | complex | str | bytes] | SeedSequence | BitGenerator | Generator | None"; expected "_SupportsArray[dtype[bool_ | integer[Any]]] | _NestedSequence[_SupportsArray[dtype[bool_ | integer[Any]]]] | bool | int | _NestedSequence[bool | int] | SeedSequence | BitGenerator | Generator | None"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

Should NumPy not expose type declarations used in the public API for re-use by other projects?

Reply:
+1 for adding type annotations.

That has a syntax error (`type Seed`) and a mistake (`rng =` is wrong). The issue is with array-likes (as usual). This works:
```python
import numpy as np
import numpy.typing

# leaving out np.typing.ArrayLike
Seed = None | int | np.random.SeedSequence | np.random.BitGenerator | np.random.Generator | np.typing.NDArray


def foo(rng : Seed = None):
    rng_ = np.random.default_rng(rng)  # note the renamed variable `rng_`

    return rng_.random((3, 3))
```

> Should NumPy not expose type declarations used in the public API for re-use by other projects?

The type annotation in NumPy for `default_rng` captures the requirement that only integer array-likes are accepted. This is correct and precise, but requires a fairly complicated annotation (look for `_ArrayLikeInt_co` in `numpy/_typing`). It does not make sense to expose such things publicly.

Array-likes in general should not be encouraged to propagate, so I'd choose to leave out `ArrayLike` completely, and replace it with `np.typing.NDArray`. That should also still gain an integer specifier: `np.typing.NDArray[np.integer]` should work.

The NumPy type annotations themselves are aiming for precision and completeness. In a SPEC, I think you want to not do that - allowing everything that numpy itself has to allow for corner-case backwards compatibility doesn't seem useful. Instead, you want to clearly capture what the function actually needs. `NDArray[np.integer]` should do that; jumping through hoops to allow arbitrary sequences of ints or even bools is counterproductive.

Mypy will be happy as long as the annotation you use is a subset of what `default_rng` accepts, rather than a superset.

Reply:
I'd also choose the type alias name for what's preferred - so either `Rng` or something like `RngOrSeed`, but not `Seed.

Reply:
Actually, lists and tuples of arbitrary-sized integers are vastly more common (and preferred) than integer arrays.

We're also missing the legacy `RandomState`, for completeness.

Reply:
> That has a syntax error (`type Seed`)

Not actually a syntax error, just newer syntax: https://peps.python.org/pep-0695/

Reply:
OK, so re-reading this thread, I'm wondering: what is the purpose of the suggested type annotation in the SPEC? Is it is to be an exhaustive description of what `default_rng` accepts, or rather to guide your library users in providing the correct arguments? I suspect we want more of the latter?

To make it very straightforward, we could then type annotate with `rng : np.random.Generator`, given that this is the usage we encourage. If we wanted to be a bit more flexible, while still guiding the user in the right direction perhaps one of:

```
type SeedOrGenerator = np.random.Generator | int | None
type SeedOrGenerator = np.random.Generator | np.random.SeedSequence | int | tuple[int] | list[int] | None
```

Or we could try and capture the whole slew of acceptable arguments, including arrays of ints, etc., but that would scarcely provide any usage guidance.

Reply:
FWIW, I think it's helpful to split this up into a `SeedLike` and an `RNGLike`.

```python
from collections.abc import Sequence
from typing import Optional

import numpy as np

# I don't know if an array of integers will typecheck as a `Sequence[int]`, but it's a rarer use case.
SeedLike = int | np.integer | Sequence[int] | np.random.SeedSequence
RNGLike = np.random.Generator | np.random.BitGenerator | np.random.RandomState
RNGOrSeed = RNGLike | SeedLike

def foo(rng: Optional[RNGOrSeed] = None):
    rng = np.random.default_rng(rng)
    ...
```

Reply:
To answer my own question from earlier: the purpose of the type annotations are to check the *validity* of calls, primarily, so we cannot use the much restricted subset. That we *can* do in the docstrings, though, with some explanatory narrative.

Reply:
@rkern I think the following will work, handling `RandomState` separately:

```python
from collections.abc import Sequence
from typing import Optional

import numpy as np


SeedLike = int | np.integer | Sequence[int] | np.random.SeedSequence
RNGLike = np.random.Generator | np.random.BitGenerator | np.random.RandomState


def foo(rng: RNGLike | SeedLike | None = None):
    if isinstance(rng, np.random.RandomState):
        rng = np.random.default_rng(rng._bit_generator)

    rng = np.random.default_rng(rng)
    ...


foo(1)
foo([1, 2, 3])
foo((1, 2, 3))
foo(np.random.default_rng(12))
```

Reply:
Maybe a bit easier to read:

```python
def ensure_rng(rng : SeedLike | RNGLike | None) -> np.random.Generator:
    if isinstance(rng, np.random.RandomState):
        return np.random.default_rng(rng._bit_generator)
    else:
        return np.random.default_rng(rng)


def foo(rng: RNGLike | SeedLike | None = None):
    rng = ensure_rng(rng)
    ...
```

Reply:
Ah, I confused myself. We've talked about making `default_rng()` accept `RandomState`s, but we never actually pushed that in. `<grumble grumble grumble>`

You can omit `RandomState`, then.

Reply:
This has now been added to the text.